File: 2015/03_Perfectly_Spherical_Houses_in_a_Vacuum/Day03.kt
// Day 3: Perfectly Spherical Houses in a Vacuum
// https://adventofcode.com/2015/day/3

import java.nio.file.Files
import java.nio.file.Paths

fun readInput(name: String): String {
    return Files.readString(Paths.get("$name.txt")).trim()
}

val DIRECTIONS = mapOf(
    '^' to Pair(0, -1),
    '>' to Pair(1, 0),
    'v' to Pair(0, 1),
    '<' to Pair(-1, 0)
)

fun part1(input: String): Int {
    var santa = Pair(0, 0)
    val visited = mutableSetOf(santa)
    for (dir in input) {
        val (dx, dy) = DIRECTIONS[dir] ?: Pair(0, 0)
        santa = Pair(santa.first + dx, santa.second + dy)
        visited.add(santa)
    }
    return visited.size
}

fun part2(input: String): Int {
    var santa = Pair(0, 0)
    var robotSanta = Pair(0, 0)
    val visited = mutableSetOf(santa, robotSanta)
    for (i in input.indices step 2) {
        val (dx1, dy1) = DIRECTIONS[input[i]] ?: Pair(0, 0)
        santa = Pair(santa.first + dx1, santa.second + dy1)
        visited.add(santa)
        if (i + 1 < input.length) {
            val (dx2, dy2) = DIRECTIONS[input[i + 1]] ?: Pair(0, 0)
            robotSanta = Pair(robotSanta.first + dx2, robotSanta.second + dy2)
            visited.add(robotSanta)
        }
    }
    return visited.size
}

fun main() {
    val day = "Day03"
    val inputTest = readInput("${day}_test")

    val input1 = ">"
    val input2 = "^>v<"
    val input3 = "^v^v^v^v^v"
    val input1v2 = "^v"

    println("Part 1: ${part1(input1)}") // 2
    println("Part 1: ${part1(input2)}") // 4
    println("Part 1: ${part1(input3)}") // 2
    println("Part 1: ${part1(inputTest)}") // 2572

    println("Part 2: ${part2(input1v2)}") // 3
    println("Part 2: ${part2(input2)}") // 3
    println("Part 2: ${part2(input3)}") // 11
    println("Part 2: ${part2(inputTest)}") // 2631
}

File: 2015/03_Perfectly_Spherical_Houses_in_a_Vacuum/Day03.js
// Day 3: Perfectly Spherical Houses in a Vacuum
// https://adventofcode.com/2015/day/3

const utils = require('../../utils');

const DIRECTIONS = {
  '^': [0, -1],
  '>': [1, 0],
  v: [0, 1],
  '<': [-1, 0],
};

function part1(input) {
  let santa = { x: 0, y: 0 };
  const visited = new Set();
  visited.add(`${santa.x},${santa.y}`);
  for (const dir of input) {
    const [dx, dy] = DIRECTIONS[dir];
    santa.x += dx;
    santa.y += dy;
    visited.add(`${santa.x},${santa.y}`);
  }
  return visited.size;
}

function part2(input) {
  let santa = { x: 0, y: 0 };
  let robotSanta = { x: 0, y: 0 };
  const visited = new Set();
  visited.add(`${santa.x},${santa.y}`);
  visited.add(`${robotSanta.x},${robotSanta.y}`);
  for (let i = 0; i < input.length; i += 2) {
    const [dx1, dy1] = DIRECTIONS[input[i]];
    const [dx2, dy2] = DIRECTIONS[input[i + 1]] ?? [0, 0];
    santa.x += dx1;
    santa.y += dy1;
    robotSanta.x += dx2;
    robotSanta.y += dy2;
    visited.add(`${santa.x},${santa.y}`);
    visited.add(`${robotSanta.x},${robotSanta.y}`);
  }
  return visited.size;
}

const day = 3;
const input1 = '>';
const input1v2 = '^v';
const input2 = '^>v<';
const input3 = '^v^v^v^v^v';
const inputTest = utils.readInputByDay(day, '_test');

console.log('Part 1:', part1(input1)); // 2
console.log('Part 1:', part1(input2)); // 4
console.log('Part 1:', part1(input3)); // 2
console.time('Part 1');
console.log('Part 1:', part1(inputTest)); // 2572
console.timeEnd('Part 1');

console.log('Part 2:', part2(input1v2)); // 3
console.log('Part 2:', part2(input2)); // 3
console.log('Part 2:', part2(input3)); // 11
console.time('Part 2');
console.log('Part 2:', part2(inputTest)); // 2631
console.timeEnd('Part 2');
